src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/ProjectSystem/DefaultWindowsRazorProjectHostTest.cs

```diff
@@ -749,8 +749,60 @@ public async Task OnProjectChanged_ReadsProperties_InitializesProject()
+    [InlineData("C:\\my repo root\\solution folder\\projectFolder\\../artifacts\\obj\\projectName\\", "C:\\my repo root\\solution folder\\projectFolder\\../artifacts\\obj\\projectName\\debug", "C:\\my repo root\\solution folder\\artifacts\\obj\\projectName\\debug")]
+    // .... and ArtifactsPivot is $(ArtifactsPivot)\_MyCustomPivot
+    [InlineData("C:\\my repo root\\solution folder\\projectFolder\\../artifacts\\obj\\projectName\\", "C:\\my repo root\\solution folder\\projectFolder\\../artifacts\\obj\\projectName\\_MyCustomPivot", "C:\\my repo root\\solution folder\\artifacts\\obj\\projectName\\_MyCustomPivot")]
+    // Set BIOP to ..\\..\\artifacts\\obj\\$(MSBuildProjectFolder), pre-ArtifactsPath existing
+    [InlineData("C:\\my repo root\\solution folder\\projectFolder\\..\\..\\artifacts\\obj\\projectName", "..\\..\\artifacts\\obj\\projectName\\Debug\\net8.0", "C:\\my repo root\\artifacts\\obj\\projectName\\Debug\\net8.0")]
+    public void IntermediateOutputPathCalculationHandlesRelativePaths(string baseIntermediateOutputPath, string intermediateOutputPath, string expectedCombinedIOP)
+    {
+
+        var services = new TestProjectSystemServices(TestProjectData.SomeProject.FilePath);
+        var host = new TestWindowsRazorProjectHost(services, _projectManagerAccessor, Dispatcher, _projectConfigurationFilePathStore, languageServerFeatureOptions: null);
+
+        var state = TestProjectRuleSnapshot.CreateProperties(
+             WindowsRazorProjectHostBase.ConfigurationGeneralSchemaName,
+             new Dictionary<string, string>()
+             {
+                 [WindowsRazorProjectHostBase.IntermediateOutputPathPropertyName] = intermediateOutputPath,
+                 [WindowsRazorProjectHostBase.BaseIntermediateOutputPathPropertyName] = baseIntermediateOutputPath,
+             });
+
+        var dict = ImmutableDictionary<string, IProjectRuleSnapshot>.Empty;
+        dict = dict.Add(WindowsRazorProjectHostBase.ConfigurationGeneralSchemaName, state);
+
+        var result = host.GetIntermediateOutputPathFromProjectChange(dict,
+            out var combinedIntermediateOutputPath);
+
+        Assert.True(result);
+        Assert.Equal(expectedCombinedIOP, combinedIntermediateOutputPath);
+    }
+
+    internal class TestWindowsRazorProjectHost : DefaultWindowsRazorProjectHost
+    {
+        public TestWindowsRazorProjectHost(IUnconfiguredProjectCommonServices commonServices,
+            IProjectSnapshotManagerAccessor projectManagerAccessor,
+            ProjectSnapshotManagerDispatcher dispatcher,
+            ProjectConfigurationFilePathStore projectConfigurationFilePathStore,
+            LanguageServerFeatureOptions languageServerFeatureOptions)
+            : base(commonServices, projectManagerAccessor, dispatcher, projectConfigurationFilePathStore, languageServerFeatureOptions)
+        {
+            SkipIntermediateOutputPathExistCheck_TestOnly = true;
+        }
+
+        // Enable access to protected method for testing
```

Comment:
We're trying to start moving to test accessors instead of exposing implementation details. Now might be a good time to do this if it's easy.

Reply:
Done